src/login/Login.h

```diff
@@ -41,19 +41,283 @@ struct ProfileRecord
+		}
+
+		return false;
+	}
+
+	UpdateCache m_update;
+	Result m_value;
+	Result m_updatedValue;
+	int m_dataVersion;
+};
+
+class StatementHelper
+{
+public:
+	StatementHelper(
+		const std::shared_ptr<WithDb>& db,
+		const std::string& query,
+		const std::function<void(sqlite3_stmt*, sqlite3*)>& bind);
+	~StatementHelper();
+
+	[[nodiscard]] bool Step() const;
+
+	template <typename T>
+	T Result(const std::function<T(sqlite3_stmt*, sqlite3*)>& result) const { return result(m_stmt, m_db); }
+
+	StatementHelper(const StatementHelper&) = delete;
+	StatementHelper(StatementHelper&&) = delete;
+	StatementHelper& operator=(const StatementHelper&) = delete;
+	StatementHelper& operator=(StatementHelper&&) = delete;
+
+private:
+	sqlite3* m_db;
+	sqlite3_stmt* m_stmt;
+};
+
+template <typename T>
+class Results
+{
+public:
+	using Type = T;
+
+	class Iterator
+	{
+	public:
+		using iterator_category = std::input_iterator_tag;
+		using difference_type = std::ptrdiff_t;
+		using value_type = T const;
+		using pointer = T const*const;
+		using reference = T const&;
+
+		explicit Iterator(Results const* results)
+			: m_done(false)
+			, m_results(results)
+		{
+			m_done = m_results->m_stmt.Step();
+			if (!m_done)
+				m_current = m_results->m_stmt.Result(m_results->m_result);
+		}
+
+		explicit operator bool() const { return !m_done; }
+
+		reference operator*() const { return m_current; }
+		pointer operator->() const { return &m_current; }
+
+		Iterator& operator++()
+		{
+			assert(!m_done);
+
+			m_done = m_results->m_stmt.Step();
+			if (!m_done)
+				m_current = m_results->m_stmt.Result(m_results->m_result);
+
+			return *this;
+		}
+
+		friend bool operator==(Iterator l, Iterator r) { return l.m_current == r.m_current; }
+		friend bool operator!=(Iterator l, Iterator r) { return !(l == r); }
+
+		class Sentinel {};
+		friend bool operator==(Iterator it, Sentinel) { return it.m_done; }
+		friend bool operator!=(Iterator it, Sentinel) { return !(it == Sentinel{}); }
+		friend bool operator==(Sentinel, Iterator it) { return it == Sentinel{}; }
+		friend bool operator!=(Sentinel, Iterator it) { return it != Sentinel{}; }
+
+	private:
+		bool m_done;
+		T m_current;
+		Results const* m_results;
+	};
+
+	explicit Results(
+		const std::shared_ptr<WithDb>& db,
+		const std::string& query,
+		const std::function<void(sqlite3_stmt*, sqlite3*)>& bind,
+		const std::function<T(sqlite3_stmt*, sqlite3*)>& result)
+		: m_db(db)
+		, m_stmt(db, query, bind)
+		, m_result(result)
+	{}
+
+	[[nodiscard]] Iterator begin() const { return Iterator(this); }
+	[[nodiscard]] typename Iterator::Sentinel end() const { return Iterator::Sentinel{}; }
+
+	[[nodiscard]] std::vector<T> vector() const
+	{
+		std::vector<T> vec; // we don't know the size, so we can't reserve
+		for (auto t : *this)
+			vec.emplace_back(std::move(t));
+		return vec;
+	}
+
+private:
+	std::shared_ptr<WithDb> m_db;
+	StatementHelper m_stmt;
+	std::function<T(sqlite3_stmt*, sqlite3*)> m_result;
+};
```

Comment:
A lot of this db helper code could probably be put into its own sqlite specific header